public/js/discover/trie.js

```diff
@@ -0,0 +1,121 @@
+const categories = [
+  'canned food',
+  'produce',
+  'cleaning supplies',
+  'blankets',
+  'clothes',
+  'pet food',
+  'plastic bags',
+  'books',
+  'school supplies',
+  "children's toys",
+  'hair products',
+  'deoderant',
+  'soap',
+  'bottled drinks',
+  'computers',
+  'cars',
+];
+
+const categoryTrie = new Trie(categories);
+
+/**
+ * Gives autocomplete suggestions for item categories.
+ * @param {string} input The prefix to base autocomplete suggestions on.
+ * @return {array} The list of autocomplete suggestions.
+ */
+window.getAutocompleteCategories = function getAutocompleteCategories(input) {
+  return categoryTrie.autocomplete(input);
+};
+
+/**
+ * The individual nodes that make up the Trie.
+ */
+class TrieNode {
+  /**
+   * Creates a node storing the given prefix.
+   * @param {string} prefix The string value at this node.
+   */
+  constructor(prefix = '') {
+    this.children = {};
+    this.isWord = false;
+    this.prefix = prefix;
+  }
+}
+
+/**
+ * A 26-ary tree for storing word prefixes and searching words.
+ */
+class Trie {
+  /**
+   * Creates a Trie based on the given words.
+   * @param {array} words The list of words to build the Trie with.
+   */
+  constructor(words) {
+    this.root = new TrieNode('');
+    for (const word of words) this.insert(word);
+  }
+
+  /**
+   * Adds the word to the Trie, creating new nodes for its prefixes if needed.
+   * @param {string} word The word to insert into the tree.
+   */
+  insert(word) {
```

Comment:
Do we remove special characters for this?

Reply:
Added. 👍